Python 3 as default python version for Fedora

mcyprian opened the pull request:
- Each disto (template) has its own default python versions
- Python 3 is now default for Fedora
- Default executables policy change

mcyprian committed:
Keep python versions in the spec always in ascending order

mcyprian committed:
Use direct formatting in template instead of macros containing hardcoded version

mcyprian committed:
Fix excluding of Sphinx build require for non-base versions

Only one version of sphinx, that will be used for sphinx build is needed,
this will exclude non-base version of sphinx if the canonical format
of the dependencies pythonXdist(sphinx) is used.

mcyprian committed:
Make Python 3 default python version for Fedora

Resolves #138

mcyprian committed:
Change executables policy

Stop creating -MAJOR.MINOR suffixed executables for each python
version. Package executables for the base python version. We expect
the same functionality for Python 2 and Python 3, which is the most
common case.

Resolves #137

mcyprian committed:
Filter __pycache__ when getting scripts from virtualenv

Python3 virtualenv is now used by default for Fedora
template so we need to filter __pycache__ as well.

irushchyshyn commented on pyp2rpm/bin.py:
```diff
@@ -80,15 +80,17 @@ def get_scl_help_record(self, ctx):
-              help='Additional Python versions to include in the specfile (e.g -p3 for %{0}).'
-              'Can be specified multiple times (default: "{1}"). Specify additional version '
+              help='Additional Python versions to include in the specfile '
+              '(e.g -p2 for {0}). Can be specified multiple times '
```
`{0}` formatting is not needed here, just do `(e.g -p2 for python2 subpackage).`

irushchyshyn commented on tests/test_data/python-sphinx.spec:
```diff
@@ -103,63 +103,29 @@ rm -rf %{pypi_name}.egg-info
-cp %{buildroot}/%{_bindir}/sphinx-quickstart %{buildroot}/%{_bindir}/sphinx-quickstart-%{python3_version}
-ln -s %{_bindir}/sphinx-quickstart-%{python3_version} %{buildroot}/%{_bindir}/sphinx-quickstart-3
-
 %py2_install
```
I really think that we should do `rm -rf %{buildroot}%{_bindir}/*` after Py2 install in such cases.

I have already seen a number of cases when the scripts did not get overwritten, resulting in a Py3 subpackage having a dependency on Py2. See the Note in the [Python RPM porting guide](https://python-rpm-porting.readthedocs.io/en/latest/application-modules.html#install). 

We could add smth like this after install:
```
{%- if data.scripts and pv != data.base_python_version %}
rm -rf %{buildroot}%{_bindir}/*
{%- endif %}
```

irushchyshyn commented on pyp2rpm/templates/fedora.spec:
```diff
@@ -47,76 +46,70 @@ rm -rf %{pypi_name}.egg-info
 {%- endfor %}
+{%- endif %}
 {%- if data.py_modules %}
 {% for module in data.py_modules -%}
```
Please use here `{%- ...` for new line control.

Otherwise it looks like this (example from Django):
```
%files -n python3-%{pypi_name}
%license LICENSE.txt
%doc DESCRIPTION.rst
%{_bindir}/django-admin.py
%{_bindir}/django-admin

%{python3_sitelib}/__pycache__/*
%{python3_sitelib}/django.py
%{python3_sitelib}/django
%{python3_sitelib}/%{pypi_name}-%{version}-py?.?.egg-info
```

irushchyshyn commented on pyp2rpm.1:
```diff
@@ -28,10 +28,10 @@ Search order is: 1) filesystem, 2) default templates.
 .TP
 .B "\-p \-\-PYTHON_VERSIONS"
-Additional Python versions to include in the specfile (e.g. -p3 f or %{?with_python3}). Specify additional version or use -b explicitly to disable default.
+Additional Python versions to include in the specfile (e.g. -p2 for python2 subpackage}). Specify additional version or use -b explicitly to disable default.
```
Please remove the closing `}`.

irushchyshyn commented on pyp2rpm/bin.py:
```diff
@@ -159,8 +161,12 @@ def main(package, v, d, s, r, proxy, srpm, p, b, o, t, venv, autonc, sclize, **s
-    if t in settings.KNOWN_DISTROS:
+    if t and os.path.splitext(t)[0] in settings.KNOWN_DISTROS:
         distro = t
+    elif t and not b or p:
```
When I do `./mybin.py -p2 Sphinx` I get:
```
Usage: mybin.py [OPTIONS] PACKAGE

Error: Default python versions for template None are missing in settings, add them or use flags -b/-p to set python versions.
```

This line should probably be changed to `elif t and not (b or p):`

irushchyshyn reviewed (commented):
That probably took a lot of work 👍 

I have added a couple of small comments and please take a look at failing travis build (the tests pass in py27 env locally).

mcyprian committed:
Run Travis CI tests in custom docker container

This will make Travis builds easily reproducible in
local environment and we will be able to test rpm
and dnf specific features.

mcyprian commented:
Travis integration test failed because metadata extraction using alternative python3 interpreter (/usr/bin/python3) didn't work in Ubuntu Trusty container used by Travis. I guess it will be better to run Travis tests inside custom Fedora container. This will make Travis builds easily reproducible in local environment in similar cases and we will be able to test rpm and dnf specific features.

irushchyshyn reviewed (approved):
Looks good!👍  Please squash the fixups and merge at will!

I would suggest though in future to make changes for separate issues in separate PRs so they are not too big.
